src/arraymancer/tensor/display.nim

```diff
@@ -13,22 +13,160 @@
 # limitations under the License.
 
 import  ./private/p_display,
-        ./data_structure,
-        typetraits
+        ./data_structure
+import std / typetraits
 
-proc pretty*[T](t: Tensor[T], precision = -1): string =
-  ## Pretty-print a Tensor with the option to set a custom `precision`
-  ## for float values.
-  var desc = t.type.name & " of shape \"" & $t.shape & "\" on backend \"" & "Cpu" & "\""
-  if t.storage.isNil: # return useful message for uninit'd tensors instead of crashing
-    return "Uninitialized " & $desc
-  if t.size() == 0:
-    return desc & "\n    [] (empty)"
-  elif t.rank == 1: # for rank 1 we want an indentation, because we have no `|`
-    return desc & "\n    " & t.prettyImpl(precision = precision)
-  else:
-    return desc & "\n" & t.prettyImpl(precision = precision)
+proc pretty*[T](t: Tensor[T], precision: int = -1, showHeader: static bool = true): string =
+  ## Pretty-print a Tensor as a "table" with a given precision and optional header
+  ##
+  ## Pretty-print a Tensor with options to set a custom `precision` for float
+  ## values and to show or hide a header describing the tensor type and shape.
+  ##
+  ## Inputs:
+  ##   - Input Tensor.
+  ##   - precision: The number of decimals printed (for float tensors),
+  ##                _including_ the decimal point.
+  ##   - showHeader: If true (the default) show a dscription header
+  ##                 indicating the tensor type and shape.
+  ## Result:
+  ##   - A string containing a "pretty-print" representation of the Tensor.
+  ##
+  ## Examples:
+  ## ```nim
+  ## let t = arange(-2.0, 4.0).reshape(2, 3)
+  ##
+  ## echo t.pretty()
+  ## # Tensor[system.float] of shape "[2, 3]" on backend "Cpu"
+  ## # |-2    -1     0|
+  ## # |1      2     3|
+  ##
+  ## # Note that the precision counts
+  ## echo t.pretty(2)
+  ## # Tensor[system.float] of shape "[2, 3]" on backend "Cpu"
+  ## # |-2.0    -1.0     0.0|
+  ## # |1.0      2.0     3.0|
+  ## ```
+  const specifier = if showHeader: "" else: "||"
+  t.prettyImpl(precision = precision, specifier = specifier)
+
+proc pretty*[T](t: Tensor[T], specifier: static string = ""): string =
+  ## Pretty-print a Tensor with the option to set a custom format `specifier`
+  ##
+  ## The "format specifier" is similar to those used in format strings, with
+  ## the addition of a few, tensor specific modifiers (shown below).
+  ##
+  ## Inputs:
+  ##   - Input Tensor
+  ##   - specifier: A format specifier similar to those used in format strings,
```

Comment:
Shouldn't those be an enum instead of static[string] ?

Reply:
Using an enum is the first thing I tried but it was not very flexible. Using a string is how I let the user configure the individual element display in the same flexible manner that nim's strformat allows. For example, you can display integers as hex or as binary values, you can force the display of the sign for positive values, you can specify the number of digits in total and on the decimal part or (with nim 2.2) you can ask to use the "math-like" "j" complex display format (which displays complex numbers as `2.3-7.1j`) which is particularly useful for a tensor library in the context of signal processing.